File: part-3-cache-mock/python/cache.py
from code import interact
from enum import Enum
from range import Range

class StorageBackend:
    """
    Define the storage backend to be plugged under the cache
    """
    def pwrite(self, data: bytearray, offset) -> int:
        return -1
    def pread(self, offset, size) -> bytearray:
        return bytearray()

class CacheEntryState(Enum):
    """
    Define the state (dirty of not) of a cache entry.
    """
    NEW = 1
    CLEAN = 2
    DIRTY = 3

class CacheEntry:
    """
    Define a cache entry which is a data segment mapped on a specific range.
    It has a clean/dirty/new state.
    """
    def __init__(self, offset, size):
        self.range = Range(offset, size)
        self.state = CacheEntryState.NEW
        self.data = bytearray(size)

class Cache:
    """
    Implement a cache which can store data an flush them to a storage backend
    On a read operation it automatically fetch the data if not already present
    in the cache
    """
    def __init__(self, backend: StorageBackend):
        self.backend = backend
        self.entries = []

    def pwrite(self, data, offset):
        # build range object to ease implementation
        op_range = Range(offset, len(data))

        # create all required new entries
        self._create_entries(op_range.offset, op_range.size)

        # loop on all
        for entry in self.entries:
            # if overlap
            if Range.overlap(entry.range, op_range):
                # calc intersect to get the copy ranges
                intersect = Range.intersect(op_range, entry.range)

                # build shifted sub-ranges in the in/out data segment.
                in_data_range = intersect.shift(-offset)
                out_data_range = intersect.shift(-entry.range.offset)

                # cpoy the data
                entry.data[out_data_range.offset:out_data_range.end()] = data[in_data_range.offset:in_data_range.end()]

                #mark dirty
                entry.state = CacheEntryState.DIRTY
    
    def pread(self, offset, size) -> bytearray:
        # build range object to ease implementation
        op_range = Range(offset, size)

        # create all required new entries
        self._create_entries(op_range.offset, op_range.size)

        # perform read operation to fetch data from the storage
        self._apply_read_on_new_entries()

        # allocate memory to load data in
        out_data = bytearray(size)

        # loop on all entries
        for entry in self.entries:
            # if overlap we need to copy
            if Range.overlap(entry.range, op_range):
                # calc intersect to find what to copy
                intersect = Range.intersect(op_range, entry.range)

                # calc buffer in & out copy ranges
                out_data_range = intersect.shift(-offset)
                in_data_range = intersect.shift(-entry.range.offset)

                # copy data
                out_data[out_data_range.offset:out_data_range.end()] = entry.data[in_data_range.offset:in_data_range.end()]
        
        # ok return the segment
        return out_data

    def flush(self, range: Range = Range(0,0)):
        # loop on all entries
        for entry in self.entries:
            # if is dirty and overlap the requested range
            if entry.state == CacheEntryState.DIRTY and (range.size == 0 or Range.collide(range, entry.range)):
                # do pwrite
                res = self.backend.pwrite(entry.data, entry.range.offset)

                # check has been done
                if res != entry.range.size:
                    raise Exception("Fail to fully write data into storage backend")

                # mark clean
                entry.state = CacheEntryState.CLEAN

    def _get_sort_key(entry: CacheEntry):
        return entry.range.offset

    def _create_entries(self, offset, size):
        # build range
        cur_range = Range(offset, size)

        # loop on all entries to find the holes we need to fill by allocating a new entry
        for i in range(len(self.entries)):
            # extract the entry
            entry = self.entries[i]

            # check if overlap
            if Range.overlap(cur_range, entry.range):
                # get the left and right segment by extruding the incomming one with the pre-existing one
                left, right = Range.exclude(cur_range, entry.range)

                # if the left residut is not NULL
                if left.size > 0:
                    self.entries.append(CacheEntry(left.offset, left.size))
                
                # replace current by the right residut
                cur_range = right
            
            # the current become NULL, we can stop we reached the enf of the overlapping region
            if cur_range.size == 0:
                break

        # There is a last one to insert after the last pre-existing one
        if cur_range.size != 0:
            self.entries.append(CacheEntry(cur_range.offset, cur_range.size))

        # sort again
        self.entries.sort(key = Cache._get_sort_key)

    def _apply_read_on_new_entries(self):
        # loop on all entries
        for entry in self.entries:
            # check if not loaded or dirty
            if entry.state == CacheEntryState.NEW:
                # load data
                data = self.backend.pread(entry.range.offset, entry.range.size)

                # failed to load
                if len(data) != entry.range.size:
                    raise Exception("Not enougth data to read")

                # fill the new entry
                entry.data = data
                entry.state = CacheEntryState.CLEAN

File: part-3-cache-mock/cpp/cache.cpp
/*********************************************************/
#include <cassert>
#include <exception>
#include <memory>
#include <cstring>
#include "cache.hpp"

/*********************************************************/
using namespace cache;

/*********************************************************/
Buffer::Buffer(size_t size)
{
	this->size = size;
	if (size == 0)
		this->data = NULL;
	else
		this->data = new char[size];
}

/*********************************************************/
Buffer::Buffer(Buffer && orig)
{
	this->size = orig.size;
	this->data = orig.data;
	orig.size = 0;
	orig.data = NULL;
}


/*********************************************************/
Buffer::Buffer(const char * value)
{
	this->size = strlen(value);
	if (this->size == 0) {
		this->data = NULL;
	} else {
		this->data = new char[this->size];
		memcpy(this->data, value, this->size);
	}
}

/*********************************************************/
Buffer::Buffer(const Buffer & orig)
{
	this->size = orig.size;
	if (this->size == 0) {
		this->data = NULL;
	} else {
		this->data = new char[this->size];
		memcpy(this->data, orig.data, this->size);
	}
}

/*********************************************************/
Buffer::Buffer(const std::string & value)
{
	this->size = value.size();
	if (this->size == 0) {
		this->data = NULL;
	} else {
		this->data = new char[this->size];
		memcpy(this->data, value.c_str(), this->size);
	}
}

/*********************************************************/
Buffer::~Buffer(void)
{
	if (this->data != NULL)
		delete [] this->data;
	this->data = NULL;
	this->size = 0;
}

/*********************************************************/
Buffer & Buffer::operator=(const Buffer & orig)
{
	if (this->size != orig.size) {
		delete [] this->data;
		this->data = new char[orig.size];
		this->size = orig.size;
	}
	memcpy(this->data, orig.data, this->size);
	return *this;
}

/*********************************************************/
bool Buffer::operator==(const Buffer & orig) const
{
	if (this->size != orig.size)
		return false;
	return strncmp(this->data, orig.data, this->size) == 0;
}

/*********************************************************/
std::ostream & cache::operator <<(std::ostream & out, const Buffer & buffer)
{
	for (size_t i = 0 ; i < buffer.size ; i++)
		out << buffer.data[i];
	return out;
}

/*********************************************************/
CacheEntry::CacheEntry(Range range)
           :data(range.get_size())
{
	this->state = CacheEntryState::NEW;
	this->range = range;
}

/*********************************************************/
CacheEntry::CacheEntry(CacheEntry && entry)
           :data(std::move(entry.data))
{
	this->state = entry.state;
	this->range = range;
}

/*********************************************************/
bool CacheEntry::operator<(const CacheEntry & other) const
{
	return this->range < other.range;
}

/*********************************************************/
Cache::Cache(StrorageBackend * backend)
{
	assert(backend != NULL);
	this->backend = backend;
}

/*********************************************************/
void Cache::create_entries(Range range)
{
	Range cur_range = range;
	for (auto & entry : this->entries) {
		if (Range::overlap(cur_range, entry.range)) {
			auto res = Range::exclude(cur_range, entry.range);
			Range left = res.first;
			Range right = res.second;
			if (left.get_size() > 0)
				this->entries.emplace_back(left);
			cur_range = right;
		}
		if (cur_range.get_size() == 0)
			break;
	}
	if (cur_range.get_size() != 0)
		this->entries.emplace_back(cur_range);
	this->entries.sort();
}

/*********************************************************/
void Cache::apply_read_on_new_entries(void)
{
	for(auto & entry : this->entries) {
		if (entry.state == CacheEntryState::NEW) {
			Buffer data = this->backend->pread(entry.range.get_offset(), entry.range.get_size());
			if (data.size != entry.range.get_size())
				throw std::exception();
			entry.data = data;
			entry.state = CacheEntryState::CLEAN;
		}
	}
}

/*********************************************************/
Buffer Cache::pread(size_t size, size_t offset)
{
	Range op_range(offset, size);
	this->create_entries(op_range);
	this->apply_read_on_new_entries();
	Buffer res(size);
	for (auto & entry : this->entries) {
		if (Range::overlap(entry.range, op_range)) {
			Range intersect = Range::intersect(op_range, entry.range);
			Range out_data_range(intersect.get_offset() - offset, intersect.get_size());
			Range in_data_range(intersect.get_offset() - entry.range.get_offset(), intersect.get_size());
			memcpy(res.data + out_data_range.get_offset(), entry.data.data + in_data_range.get_offset(), out_data_range.get_size());
		}
	}
	return res;
}

/*********************************************************/
size_t Cache::pwrite(const Buffer & value, size_t offset)
{
	Range op_range(offset, value.size);
	this->create_entries(op_range);
	for (auto & entry : this->entries) {
		if (Range::overlap(entry.range, op_range)) {
			Range intersect = Range::intersect(op_range, entry.range);
			Range out_data_range(intersect.get_offset() - offset, intersect.get_size());
			Range in_data_range(intersect.get_offset() - entry.range.get_offset(), intersect.get_size());
			memcpy(entry.data.data + in_data_range.get_offset(), value.data + out_data_range.get_offset(), out_data_range.get_size());
			entry.state = CacheEntryState::DIRTY;
		}
	}
	return value.size;
}

/*********************************************************/
void Cache::flush(Range range)
{
	for (auto & entry : this->entries) {
		if (entry.state == CacheEntryState::DIRTY && (range.get_size() == 0 || Range::overlap(range, entry.range))) {
			size_t res = this->backend->pwrite(entry.data, entry.range.get_offset());
			if (res != entry.range.get_size())
				throw std::exception();
			entry.state = CacheEntryState::CLEAN;
		}
	}
}
